streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -237,10 +247,79 @@ public final TopologyBuilder addProcessor(String name, ProcessorSupplier supplie
 
         nodeNames.add(name);
         nodeFactories.add(new ProcessorNodeFactory(name, parentNames, supplier));
+        nodeGroups.add(name);
+        nodeGroups.unite(name, parentNames);
         return this;
     }
 
     /**
+     * Returns the topic groups.
+     * A topic group is a group of topics in the same task.
+     *
+     * @return groups of topic names
+     */
+    public Collection<Set<String>> topicGroups() {
+        List<Set<String>> topicGroups = new ArrayList<>();
+
+        for (Set<String> nodeGroup : generateNodeGroups(nodeGroups)) {
+            Set<String> topicGroup = new HashSet<>();
+            for (String node : nodeGroup) {
+                String[] topics = nodeToTopics.get(node);
+                if (topics != null)
+                    topicGroup.addAll(Arrays.asList(topics));
+            }
+            topicGroups.add(Collections.unmodifiableSet(topicGroup));
+        }
+
+        return Collections.unmodifiableList(topicGroups);
+    }
+
+    private Collection<Set<String>> generateNodeGroups(QuickUnion<String> grouping) {
+        HashMap<String, Set<String>> nodeGroupMap = new HashMap<>();
+
+        for (String nodeName : nodeNames) {
+            String root = grouping.root(nodeName);
+            Set<String> nodeGroup = nodeGroupMap.get(root);
+            if (nodeGroup == null) {
+                nodeGroup = new HashSet<>();
+                nodeGroupMap.put(root, nodeGroup);
+            }
+            nodeGroup.add(nodeName);
+        }
+
+        return nodeGroupMap.values();
+    }
+
+    /**
+     * Asserts that the streams of the specified source nodes must be copartitioned.
+     *
+     * @param sourceNodes a set of source node names
+     */
+    public void copartitionSources(Collection<String> sourceNodes) {
+        copartitionSourceGroups.add(Collections.unmodifiableSet(new HashSet<>(sourceNodes)));
+    }
+
+    /**
+     * Returns the copartition groups.
+     * A copartition group is a group of topics that are required to be copartitioned.
+     *
+     * @return groups of topic names
+     */
+    public Collection<Set<String>> copartitionGroups() {
```

Comment:
If there is a topology with (A join B) join C, then ```copartitionSourceGroups``` will have two sets: {A, B} and {A, B, C}. Is it OK?